drivers/sensor/s11059/s11059.c

```diff
@@ -0,0 +1,304 @@
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(S11059, CONFIG_SENSOR_LOG_LEVEL);
+
+/* register address */
+#define S11059_REG_ADDR_CONTROL	      0x00
+#define S11059_REG_ADDR_MANUAL_TIMING 0x01
+#define S11059_REG_ADDR_DATA	      0x03
+
+/* control bit */
+#define S11059_CONTROL_GAIN	       3
+#define S11059_CONTROL_STANDBY_MONITOR 5
+#define S11059_CONTROL_STADBY	       6
+#define S11059_CONTROL_ADC_RESET       7
+
+/* bit mask for control */
+#define S11059_BIT_MASK_INTEGRATION_TIME	0x03
+#define S11059_BIT_MASK_CONTROL_STANDBY_MONITOR 0x20
+
+/* factors for converting sensor samples to Lux */
+#define S11059_CONVERT_FACTOR_LOW_RED	 (112)
+#define S11059_CONVERT_FACTOR_LOW_GREEN	 (83)
+#define S11059_CONVERT_FACTOR_LOW_BLUE	 (44)
+#define S11059_CONVERT_FACTOR_LOW_IR	 (3 * 10)
+#define S11059_CONVERT_FACTOR_HIGH_RED	 (117 * 10)
+#define S11059_CONVERT_FACTOR_HIGH_GREEN (85 * 10)
+#define S11059_CONVERT_FACTOR_HIGH_BLUE	 (448)
+#define S11059_CONVERT_FACTOR_HIGH_IR	 (30 * 10)
+
+#define S11059_INTEGRATION_TIME_MODE_00 175
+#define S11059_INTEGRATION_TIME_MODE_01 2800
+#define S11059_INTEGRATION_TIME_MODE_10 44800
+#define S11059_INTEGRATION_TIME_MODE_11 358400
+
+#define S11059_WAIT_PER_LOOP	 400
+#define S11059_INITIAL_CONTROL	 0x04
+#define S11059_MAX_MANUAL_TIMING UINT16_MAX
+#define S11059_CARRY_UP		 10000
+
+#define S11059_NUM_GAIN_MODE 2
+
+enum s11059_channel {
+	RED,
+	GREEN,
+	BLUE,
+	IR,
+	NUM_OF_COLOR_CHANNELS
+};
+
+struct s11059_dev_config {
+	struct i2c_dt_spec bus;
+	uint8_t gain;
+	int64_t integration_time; /* integration period (unit: us) */
+};
+
+struct s11059_data {
+	uint16_t samples[NUM_OF_COLOR_CHANNELS];
+};
+
+static const uint16_t convert_factors[S11059_NUM_GAIN_MODE][NUM_OF_COLOR_CHANNELS] = {
+	{S11059_CONVERT_FACTOR_LOW_RED, S11059_CONVERT_FACTOR_LOW_GREEN,
+	 S11059_CONVERT_FACTOR_LOW_BLUE, S11059_CONVERT_FACTOR_LOW_IR},
+	{S11059_CONVERT_FACTOR_HIGH_RED, S11059_CONVERT_FACTOR_HIGH_GREEN,
+	 S11059_CONVERT_FACTOR_HIGH_BLUE, S11059_CONVERT_FACTOR_HIGH_IR}};
+
+/* Integration timing in Manual integration mode */
+static const uint32_t integ_time_factor[] = {
+	S11059_INTEGRATION_TIME_MODE_00, S11059_INTEGRATION_TIME_MODE_01,
+	S11059_INTEGRATION_TIME_MODE_10, S11059_INTEGRATION_TIME_MODE_11};
+
+static int s11059_convert_channel_to_index(enum sensor_channel chan)
+{
+	switch (chan) {
+	case SENSOR_CHAN_RED:
+		return RED;
+	case SENSOR_CHAN_GREEN:
+		return GREEN;
+	case SENSOR_CHAN_BLUE:
+		return BLUE;
+	default:
+		return IR;
+	}
+}
+
+static int s11059_samples_read(const struct device *dev, uint8_t addr, uint16_t *val, uint32_t size)
+{
+	const struct s11059_dev_config *cfg = dev->config;
+	int rc;
+
+	rc = i2c_burst_read_dt(&cfg->bus, addr, (uint8_t *)val, size);
+	if (rc < 0) {
+		return rc;
+	}
+
+	for (size_t i = 0; i < NUM_OF_COLOR_CHANNELS; i++) {
+		val[i] = sys_be16_to_cpu(val[i]);
+	}
```

Comment:
This will overrun the buffer if size < NUM_OF_COLOR_CHANNELS*2

Reply:
Added processing to except an invalid `size` to avoid the buffer overrun.

https://github.com/hirokin94/zephyr/blob/d99ea531832245f93bf15b8473959fb4c9ee3de5/drivers/sensor/s11059/s11059.c#L107-L109